lms/static/scripts/frontend_apps/utils/api.js

```diff
@@ -1,15 +1,69 @@
 import { APIError } from '../errors';
 
+/**
+ * Error response when a call fails due to expiry of an access token for an
+ * external API.
+ *
+ * @typedef RefreshError
+ * @prop {RefreshCall} refresh
+ */
+
+/**
+ * Parameters for an API call that will refresh an expired access token.
+ *
+ * @typedef RefreshCall
+ * @prop {string} method
+ * @prop {string} path
+ */
+
+/**
+ * Check if an API error response indicates that a token refresh is needed.
+ *
+ * @param {any} data - Parsed body of an API error response
+ * @return {data is RefreshError}
+ */
+function isRefreshError(data) {
+  return data && data.refresh && typeof data.refresh === 'object';
+}
+
+/**
+ * Map of request path to result promise for access token refresh requests that are
+ * in flight.
+ *
+ * This is used to avoid triggering concurrent refresh requests for the same
+ * external API.
+ *
+ * @type {Map<string, Promise<void>>}
+ */
+const activeRefreshCalls = new Map();
+
 /**
  * Make an API call to the LMS app backend.
  *
  * @param {object} options
+ *   @param {string} options.authToken - Session authorization token
  *   @param {string} options.path - The `/api/...` path of the endpoint to call
- *   @param {string} options.authToken
+ *   @param {boolean} [options.allowRefresh] - If the request fails due to
+ *     an expired access token for an external API, this flag specifies whether
+ *     to attempt to refresh the token.
  *   @param {object} [options.data] - JSON-serializable body of the request
+ *   @param {string} [options.method] - Custom HTTP method for call. Defaults
+ *     to GET, or POST if `data` is set.
  *   @param {Record<string, string>} [options.params] - Query parameters
+ * @return {Promise<any>} - Parsed JSON response. TODO: Convert this to `Promise<unknown>`
```

Comment:
`Promise<unknown>` can't be used here now because...?

Reply:
When we do that, all the places in the code that call this function will need to explicitly cast the result to the structure they expect the call to return. I think that's out of scope for this PR.